libpermian/plugins/anaconda_webui/hypervisor.py

```diff
@@ -0,0 +1,140 @@
+        :rtype: list
+        """
+        if not self.remote:
+            LOGGER.info('Forwarding is not needed on local hypervisor')
+            return ports
+
+        return [ self._forward_port(vm_ip, port) for port in ports ]
+
+    def _forward_port(self, target_ip, target_port):
+        for _ in range(100):
+            local_port = random.randint(49152, 65535)
+            try:
+                subprocess.run(
+                    ['ssh', '-S', self.ssh_ctl_file,
+                    '-O', 'forward',
+                    '-L', f'*:{local_port}:{target_ip}:{target_port}',
+                    f'{self.host}'
+                    ], check=True)
+                LOGGER.debug(f'Forwarded {local_port} to {target_ip}:{target_port}')
+                return local_port
+            except:
+                # couldn't use local_port, let's try another one
+                continue
+
+    def wait_for_ip(self, vm_name, attempts=30, wait=10):
+        """  Wait for VM to get IP
+
+        :param vm_name: VM name
+        :type vm_name: string
+        :param attempts: Number of attempts
+        :type attempts: integer
+        :param wait: How log to wait before next attempt (seconds)
+        :type wait: float
+        :return: CompletedProcess
+        :rtype: string
+        """
+        for _ in range(attempts):
+            time.sleep(wait)
+            output = self.virsh_call(['domifaddr', vm_name])
+            if output:
+                return output.split()[3].split('/')[0]
+        else:
+            raise Exception('Timeout, VM still doesn\'t have IP')
+
+    def stop_vm(self, vm_name):
+        """ Call destroy on specified VM """
+        self.virsh_call(['destroy', vm_name])
+
+    def remove_vm(self, vm_name):
+        """ Remove specified VM including storage """
+        try:
+            self.virsh_call(['undefine', vm_name, '--remove-all-storage'], check=True)
+        except subprocess.CalledProcessError:
+            LOGGER.error(f'Failed to undefine VM {vm_name}')
+
+    def virsh_call(self, args, check=False):
+        """ Runs virsh with specified arguments
+
+        :param args: virsh arguments
+        :type args: list
+        :param check: Check if command was successful, defaults to False
+        :type check: bool, optional
+        :return: CompletedProcess
+        :rtype: subprocess.CompletedProcess
+        """
+        cmd = ['virsh', '-q', '--connect', self.qemu_host] + args
+        LOGGER.debug('Running: ' + repr(cmd))
+        return subprocess.run(cmd, check=check, stdout=subprocess.PIPE).stdout.decode()
+
+    def _run_cmd(self, cmd):
+        """ Run command on hypervisor
+
+        :param cmd: Command
+        :type cmd: string
+        :return: Command output, including stderr
+        :rtype: string
+        """
+        proc = subprocess.run(['ssh', '-p', '22', f'{self.host}', cmd], stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```

Comment:
Since there's already ssh connection using `self.ssh_ctl_file`, there should be no need to create a new ssh connection for command execution.

Reply:
_run_cmd is not used anymore, I will remove this function